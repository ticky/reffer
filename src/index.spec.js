/* global describe, it, expect */
import reffer from './';

describe('Reffer', () => {
  describe('with Babel function bind syntax', () => {
    it('returns a function', () => {
      expect(() => ({}::reffer('meow'))).not.toThrow();
      expect({}::reffer('meow')).toBeInstanceOf(Function);
    });

    it('returns the same function as `.ref`', () => {
      const returnVal = {}::reffer('meow');
      expect(returnVal).toBeInstanceOf(Function);
      expect(returnVal.ref).toBeInstanceOf(Function);
      expect(returnVal).toBe(returnVal.ref);
    });

    describe('callback function', () => {
      it('applies the second supplied argument to the specified key on the object', () => {
        const myObject = {};
        const callback = myObject::reffer('nya');
        callback('purr');
        expect(myObject['nya']).toBe('purr');
      });
    });
  });

  describe('with ES5-compatible syntax', () => {
    it('returns a function', () => {
      expect(() => reffer({}, 'meow')).not.toThrow();
      expect(reffer({}, 'meow')).toBeInstanceOf(Function);
    });

    it('returns the same function as `.ref`', () => {
      const returnVal = reffer({}, 'meow');
      expect(returnVal).toBeInstanceOf(Function);
      expect(returnVal.ref).toBeInstanceOf(Function);
      expect(returnVal).toBe(returnVal.ref);
    });

    describe('callback function', () => {
      it('applies the second supplied argument to the specified key on the object', () => {
        const myObject = {};
        const callback = reffer(myObject, 'nya');
        callback('purr');
        expect(myObject['nya']).toBe('purr');
      });
    });
  });

  it('throws an error if given an invalid argument list', () => {
    expect(() => reffer({}, 'meow', 'purr')).toThrowErrorMatchingSnapshot();
    expect(() => reffer()).toThrowErrorMatchingSnapshot();
    expect(() => ({}::reffer())).toThrowErrorMatchingSnapshot();
  });
});
