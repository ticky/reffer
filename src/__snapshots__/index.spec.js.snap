exports[`Reffer throws an error if given an invalid argument list 1`] = `"reffer: must be supplied with either 1 or 2 arguments. Got 3."`;

exports[`Reffer throws an error if given an invalid argument list 2`] = `"reffer: must be supplied with either 1 or 2 arguments. Got 0."`;

exports[`Reffer throws an error if given an invalid argument list 3`] = `"reffer: must be supplied with either 1 or 2 arguments. Got 0."`;
